### PR TITLE
Random delay for opennet ack when not path folding (locally)

### PR DESCRIPTION
**Pull request by toad**
This should improve protection against local attackers. I would like it to be considered for 1472 if not earlier. The performance impact should be marginal: Path folding does not block returning the data, although maybe we might end up running slightly fewer requests as a result. If so increasing the limits (BANDWIDTH_LIABILITY_*) in NodeStats would be a simple solution.

### CODE REVIEW

**Comment by toad** [comment]
Also, if the delays are too long this might impact on newbie nodes. We might want to adjust the constants, or reconsider whether to include both bwlimitDelayTime and median ping time etc. Currently my big node is showing opennet delays of around 5 seconds.

**Comment by toad** [comment]
More like 2 seconds now with some tweaking, but still looks valid IMHO.

File: src/freenet/node/RequestSender.java [hunk 1]
```diff
@@ -1681,7 +1699,7 @@ void ackOpennet(final PeerNode next) {
      * Add the peer.
      * @return True only if there was a fatal timeout and the caller should not unlock.
      */
-    private boolean finishOpennet(PeerNode next) {
+    private boolean finishOpennet(final PeerNode next) {
```
Comment (ArneBab) [inline src/freenet/node/RequestSender.java]: is this change related to the pull-request?

Comment (toad) [inline src/freenet/node/RequestSender.java]: Yes, you can't access non-final local variables from anonymous inner classes. We need this variable in run() below.

File: src/freenet/node/RequestSender.java [hunk 2]
```diff
@@ -1716,9 +1734,18 @@ private boolean finishOpennet(PeerNode next) {
...
+				    if(logMINOR) Logger.minor(this, "Delaying opennet completion for "+TimeUtil.formatTime(delay, 2, true));
+				    node.ticker.queueTimedJob(new Runnable() {
+
+                        @Override
+                        public void run() {
+                            ackOpennet(next);
+                        }
+				        
+				    }, delay);
+				} else if(origTag.shouldStop()) {
```
Comment (toad) [inline src/freenet/node/RequestSender.java]: This is the important bit: We call ackOpennet() after a computed delay, not immediately.

File: src/freenet/node/RequestSender.java (hunk 1 above)
Comment (ArneBab) [inline src/freenet/node/RequestSender.java]: ok, thanks for the explanation!

**Comment by bertm** [comment]
I don't think this one and #490 are mutually exclusive, do you @toad?

**Comment by toad** [comment]
This only happens at the originator, i.e. at maximum HTL. So it would be meaningless to have both. We could do the randomisation when we start path folding, but there wouldn't be much point given HTL gives it away anyway so it's not revealing anything useful.

**Comment by toad** [comment]
That is, I don't think we need both. But I think this is less risky than #490.

**Comment by ArneBab** [comment]
Not path folding at high HTL is also useful to keep the self-optimization working if we have to short paths (as we currently do), because that makes the weight of local requests far too big compared to remote requests.

#490 isn’t only for security, but also for improving path folding.
